### PR TITLE
sdk: Fixes to `Room::forget`

### LINKED ISSUE #4078: Calling `Room::forget` should remove the room from `m.direct` account data.
This was brought up in https://github.com/matrix-org/matrix-rust-sdk/pull/4074#issuecomment-2396156907. According to the spec, the homeserver has no obligation to keep the `m.direct` account data up-to-date. That means that we can have the `m.direct` account data with rooms that are unknown because they were left then forgotten.

To fix that we should remove the room from the `m.direct` account data when `Room::forget` is called.

### LINKED ISSUE #4079: Calling `Room::forget` should remove room from in-memory map in `Store`
`Room::forget` calls `store.remove_room()`, but since this is the method from the `StateStore` trait, the room is only removed from the database, not from the in-memory `Store::rooms` map.

There should maybe be a wrapper method on `Store` that removes it both from the database and the in-memory map.

### PR DESCRIPTION
**Pull request by zecakeh**
When forgetting rooms:
- Remove direct rooms from the `m.direct` account data,
- Remove rooms from the in-memory list.

Fixes #4078.\
Fixes #4079.

### CODE REVIEW

File: crates/matrix-sdk-base/src/client.rs [hunk 1]
```diff
@@ -1403,6 +1403,15 @@ impl BaseClient {
         self.store.room(room_id)
     }
 
+    /// Forget the room with the given room ID.
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/client.rs:1406]: Might be useful to explain how it differs from leaving a room?

File: crates/matrix-sdk-base/src/store/observable_map.rs [hunk 2]
```diff
@@ -249,6 +272,30 @@ mod tests {
...
+        assert_eq!(map.get(&'a'), Some(&'e'));
+        assert_eq!(map.get(&'b'), Some(&'f'));
+        assert!(map.get(&'c').is_none());
+
+        // remove one item
+        map.remove(&'b');
+
+        assert_eq!(map.get(&'a'), Some(&'e'));
+        assert_eq!(map.get(&'b'), None);
+        assert_eq!(map.get(&'c'), None);
```
Comment (Hywan) [inline crates/matrix-sdk-base/src/store/observable_map.rs:296]: Can you also `map.remove(&'d')` —which doesn't exist— to assert it returns `None` please?

**Review by Hywan — CHANGES_REQUESTED** [review]
It looks good, as always, thanks!

A small feedback about tests, otherwise it's good.

File: crates/matrix-sdk-base/src/store/observable_map.rs (hunk 2 above)
Comment (zecakeh) [reply crates/matrix-sdk-base/src/store/observable_map.rs:296]: Ah good catch!

File: crates/matrix-sdk-base/src/client.rs (hunk 1 above)
Comment (zecakeh) [reply crates/matrix-sdk-base/src/client.rs:1406]: Is what I added okay?

Comment (Hywan) [reply crates/matrix-sdk-base/src/client.rs:1406]: But it's sending particular request. I think a user may not want to read the spec to understand the difference between `room.leave()` and `room.forget()`. Can we clarify that please?

File: crates/matrix-sdk-base/src/store/observable_map.rs (hunk 2 above)
Comment (Hywan) [reply crates/matrix-sdk-base/src/store/observable_map.rs]: Nice, thanks!

File: crates/matrix-sdk-base/src/client.rs (hunk 1 above)
Comment (zecakeh) [reply crates/matrix-sdk-base/src/client.rs:1406]: This is the base SDK, there is not request.

**Review by poljar — APPROVED** [review]
I think that this looks fine, can you just get rid of the fixup commit?

**Comment by zecakeh** [comment]
Done.

**Pull request merged**